### DISCUSSION TITLE
Removal of ResultProxy in 1.4 is a breaking change

### DISCUSSION THREAD

**Discussion by thomasfulton** [sqlalchemy/sqlalchemy#6119] [Ideas]
Classes shouldn't be removed between minor versions as it has the potential to spontaneously break projects (e.g. mine)

**Comment by sigmavirus24** [comment]
The class still appears to be [in the library](https://github.com/sqlalchemy/sqlalchemy/blob/a22b4dfbc44c78cb4fce27c33cf74cf57878a3b1/lib/sqlalchemy/engine/cursor.py#L1836) and the new and better [objects claim backwards compatibility](https://github.com/sqlalchemy/sqlalchemy/blob/fb9099d8c4b0bca350d1a88e22fcfb6a8f2d079d/doc/build/changelog/migration_14.rst#new-result-object) so I'm not sure this issue, as written is valid.

**Reply by CaselIT** [reply]
I think ResultProxy is no longer exported in the [engine package](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/__init__.py) so the import will break.
Still 1.3->1.4 a major version upgrade for sqlalchemy

**Reply by zzzeek** [reply]
well that's a bug because we have the other ResultProxy names in there

**Reply by CaselIT** [reply]
should be maybe warn for python 3.7+ on the import?

**Reply by zzzeek** [reply]
I'm not even sure how to do that.   there's not much reason if any to actually import ResultProxy IMO.

**Reply by CaselIT** [reply]
I think it's mainly import for pep484, at least I did have one import break and I used it for that reason.

Bet py3.7+ supprots module `__getattr__`. Example: https://github.com/pandas-dev/pandas/blob/3b60759002fdc13edbbb1eab6c4bc5ec14d027ef/pandas/__init__.py#L193-L245

**Reply by zzzeek** [reply]
too much for right now.   1.4 is tired and needs to rest

**Reply by CaselIT** [reply]
I guess we can decide down the road what would be deprecated for 2.0 and maybe introduce that import patter then in a point release

**Comment by CaselIT** [comment]
Hi,

Sqlalchemy does not follow semver, so the change 1.3 -> 1.4 is a major version in sqlalchemy, as it has always being.
Changes in the same series 1.3.x and 1.4.x are in general compatible between them, but there may be some small incompatible changes in some cases.

**Comment by zzzeek** [comment]
we have a detailed document describing our versioning procedure, which is not semver right now, at https://www.sqlalchemy.org/download.html#versions

**Reply by CaselIT** [reply]
I forgot about it :)

**Comment by zzzeek** [comment]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2668
